### PR TITLE
Add GFX to UBL mesh edit screen

### PR DESCRIPTION
**Pull request by cdedwards**
This adds the up/down and rotation graphics to the mesh editing screen of UBL.
Shamelessly stolen from  marcio-ao

### CODE REVIEW

**Comment by Roxy-3D** [comment]
@cdedwards    Probably we should make a small change.  It would be good to add a SanityCheck.h condition to look for AUTO_BED_LEVELING_UBL_GFX being enabled with a non-graphics LCD Panel.

It maybe we can define similar custom characters for the 20x4 LCD Panels and do the same thing on those displays.   That would be really nice.    During the duration of the Mesh Edit screen, I don't think any custom characters are in use.   So it might be possible to use several of them to form the circles and arrows.

UPDATE:  Travis is blowing up with the G3D_PANEL displays.  That is a 20x4 LCD Panel.

@Tannoo   Do you have any comments how this functionality can be replicated on the 20x4 LCD's ?

@thinkyhead  How disruptive is it to merge stuff into bugfix-v1.1.x right now?  I'm guessing we should wait until the new organization can pass Travis and then put this into the new v2.0.0 branch?

**Comment by marcio-ao** [comment]
@cdedwards: It seems like you are duplicating the graphics I already added to "[dogm_bitmaps.h](https://github.com/MarlinFirmware/Marlin/blob/bugfix-1.1.x/Marlin/dogm_bitmaps.h#L423)". Why don't you use the same bitmaps and simply add an OR clause to the #define so those graphics can be used by either feature? Also, [_lcd_babystep_zoffset_overlay](https://github.com/MarlinFirmware/Marlin/blob/386b74f8c253d912b96064fdeb5250cf665527ca/Marlin/ultralcd.cpp#L1002) could probably be renamed to something more generic and reused.

**Comment by Roxy-3D** [comment]
I just loaded this and tried it out.  I have a question:    In the middle top of the LCD Panel there is a line.   Some times it has a single dot above it and other times nothing.   Is it supposed to be drawing the nozzle getting closer (or further) from the line?   

With the gMax configuration files...   I'm just getting a dot or no dot.

**Comment by marcio-ao** [comment]
@Roxy-3D: Since this is based on my original code, let me chime in. My guess is that the following lines need to be moved inside the if statement, after "lcd_implementation_drawedit".

```
#if ENABLED(AUTO_BED_LEVELING_UBL_GFX)
      _lcd_ubl_zoffset_overlay(mesh_edit_value);
#endif
```

File: Marlin/ultralcd.cpp [hunk 1]
```diff
@@ -1077,6 +1077,137 @@ void kill_screen(const char* lcd_msg) {
   #endif // BABYSTEPPING
 
   #if ENABLED(AUTO_BED_LEVELING_UBL)
+    #if ENABLED(AUTO_BED_LEVELING_UBL_GFX)
```
Comment (marcio-ao) [inline Marlin/ultralcd.cpp]: This should use the same graphics as BABYSTEP_ZPROBE_GFX_OVERLAY, as duplicating them would use up a lot of memory. 

Where I have "#if ENABLED(BABYSTEP_ZPROBE_GFX_OVERLAY)", simply use "#if ENABLED(BABYSTEP_ZPROBE_GFX_OVERLAY) || ENABLED(AUTO_BED_LEVELING_UBL)

**Comment by Roxy-3D** [comment]
I'm kind of guessing what you are thinking.   Do you mean move the order of the lines to be:
```cpp
      if (lcdDrawUpdate)
        lcd_implementation_drawedit(msg, ftostr43sign(mesh_edit_value));
    }
#if ENABLED(AUTO_BED_LEVELING_UBL_GFX)
      _lcd_ubl_zoffset_overlay(mesh_edit_value);
#endif
```

File: Marlin/ultralcd.cpp [hunk 2]
```diff
@@ -1077,6 +1077,137 @@ void kill_screen(const char* lcd_msg) {
...
+      0x7F,0x80, // 0111111110000000
+      0xFF,0xC0, // 1111111111000000
+      0xFF,0xC0, // 1111111111000000
+      0xFF,0xC0, // 1111111111000000
+      0x3F,0x00, // 0011111100000000
+      0x1E,0x00, // 0001111000000000
+      0x0C,0x00  // 0000110000000000
+    };
+
+        void _lcd_ubl_zoffset_overlay(float zprobe_zoffset) {
```
Comment (marcio-ao) [inline Marlin/ultralcd.cpp]: I think this function could be shared between BABYSTEP_ZPROBE_GFX_OVERLAY and AUTO_BED_LEVELING_UBL. Although they would end up sharing a static variable for keeping track of turn direction, it should still work ok as the graphics would be updated as soon as the user started to turn the dial.

File: Marlin/ultralcd.cpp [hunk 3]
```diff
@@ -1096,7 +1227,9 @@ void kill_screen(const char* lcd_msg) {
         const int32_t rounded = (int32_t)(mesh_edit_value * 1000.0);
         mesh_edit_value = float(rounded - (rounded % 5L)) / 1000.0;
       }
-
+#if ENABLED(AUTO_BED_LEVELING_UBL_GFX)
```
Comment (marcio-ao) [inline Marlin/ultralcd.cpp]: This needs to be inside the "if(lcdDrawUpdate)", after "lcd_implementation_drawedit". The reason being that if it isn't, the graphics may be overwritten.

**Review by marcio-ao — CHANGES_REQUESTED** [review]
A few suggestions.

**Comment by marcio-ao** [comment]
@roxy-3d: No, inside the "if":

```
    if (lcdDrawUpdate)
        lcd_implementation_drawedit(msg, ftostr43sign(mesh_edit_value));
        #if ENABLED(AUTO_BED_LEVELING_UBL_GFX)
           _lcd_ubl_zoffset_overlay(mesh_edit_value); 
        #endif
    }
```

The reason is that "lcd_implementation_drawedit" is what draws the screen and only when that is called we want to update the graphical overlay (which is drawn over the regular screen).

**Comment by Roxy-3D** [comment]
@marcio-ao    @cdedwards  Nope...   Something else will need to be done.     I'm not sure if the change you suggested did this...  But now the arrows move up and down a little bit next to the clockwise and counter clockwise circles.   But the nozzle is still not visible in relation to the line.

Also...   Do we really need a graphic defined for the nozzle?   In the vanilla Graphical LCD Panel, there is a nozzle defined for the heater temperature.   Would that nozzle graphic be suitable for this purpose?

**Comment by marcio-ao** [comment]
@Roxy-3D: I'm stumped then. Does the BABYSTEP_ZPROBE_GFX_OVERLAY work for you? cdedwards patch is based on that.

**Comment by Roxy-3D** [comment]
When I go to Z Baby Stepping...  I just have the standard text line.  No graphics there.

**Comment by marcio-ao** [comment]
@Roxy-3D: Well, you need to enable the following:

- BABYSTEPPING
- BABYSTEP_ZPROBE_OFFSET
- BABYSTEP_ZPROBE_GFX_OVERLAY

In "Configuration_adv.h"

**Comment by Roxy-3D** [comment]
I don't want the BABYSTEP_ZPROBE_OFFSET enabled....  (I'm using UBL and that option doesn't really make sense there.)    But BABYSTEPPING is turned on, and I just turned on BABYSTEP_ZPROBE_GFX_OVERLAY.    SanityCheck.h complains I need to have BABYSTEP_ZPROBE_OFFSET turned on to use the BABYSTEP_ZPROBE_GFX_OVERLAY option.

I guess you know that!  But that doesn't seem right to me.   It seems like the graphics should be able to show up even if the Z_PROBE_OFFSET_FROM_EXTRUDER number is not being messed with.   (I very much do not want it changed when I use Z-BabyStepping.)

It would seem there shouldn't be separate options.    It would seem to me we should just have one option controlling the Graphics overlays.   Would it make sense to combine AUTO_BED_LEVELING_UBL_GFX and BABYSTEP_ZPROBE_GFX_OVERLAY to control things?

I guess what I'm really asking is does the user need that much configuration granularity?  If they turn on the graphics overlay in one place...  Isn't it very likely they want it everywhere?

**Comment by marcio-ao** [comment]
@Roxy-3D: Yes, I suggested to @cdedwards that AUTO_BED_LEVELING_UBL_GFX and BABYSTEP_ZPROBE_GFX_OVERLAY should be combined. When I made my PR I was developing for the LulzBot printers, which do not use UBL. I suspect correct way to add support for a graphical overlay to UBL would be to share the code and configuration between the two.

You wouldn't want the overlay everywhere, though. The nozzle graphics only makes sense for operations that involve moving the nozzle up and down.

**Comment by Roxy-3D** [comment]
> You wouldn't want the overlay everywhere, though. The nozzle graphics only makes sense for operations that involve moving the nozzle up and down.

Right!   But there are number of places where that will help make things more obvious for the user.  G2 P4 is probably the most important.  But G29 P2 (manual probe) would be helpful.   And of course...  The Z-BabyStepping.

**Comment by marcio-ao** [comment]
@Roxy-3D: I agree. It would be fairly easy to do. The way to do it would be to take the changes I did in PR #7593 and generalize it a bit. "_lcd_babystep_zoffset_" overlay could be renamed something generic such as "_lcd_nozzle_height_adjustment_overlay" or something and then used from within all screens that involve raising or lowering the nozzle.

**Comment by marcio-ao** [comment]
@Roxy-3D: I suggest integrating PR #7602 first though. @GMagician found a small bug in my PR #7593

**Comment by Roxy-3D** [comment]
Yeah.  Ok!   Probably what makes sense to do is wait a few days until the v2.0.0 stuff shows up with the new file and folder layout.   And then move all these various changes to that code base.    The problem is merging stuff into bugfix-v1.1.x is disruptive...   And besides that...  It probably gets left behind when the v2.0.0 branch appears.

**Comment by marcio-ao** [comment]
@Roxy-3D: Does 2.0.0 have any changes other than file location changes? Should I be creating PRs for 2.0 rather than bugfix-v1.1.x?

**Comment by marcio-ao** [comment]
@Roxy-3D: Although I agree we may want to put off adding the overlay to other screens right now, I would suggest merging PR #7602 as my earlier PR #7593 has *already* been merged, so it's only a bug fix to something that has already been accepted.

**Comment by Roxy-3D** [comment]
Pretty much...  v2.0.0 is a version of bugfix-v1.1.x frozen a 4 or 5 days ago.   

The only changes in v2.0.0 should be what it takes to break apart the code base into smaller modules and put those files into their own folders.     You know how reliable software schedules are.   But supposedly in a few days we should be able to start using and doing Pull Requests against v2.0.0

So...  in my opinion...   The answer is "Yes...   we should be doing Pull Requests against v2.0.0 but the problem is it isn't fully baked yet and we don't really want to merge anything new into yet anyway."

**Comment by Roxy-3D** [comment]
> I would suggest merging PR #7602 as my earlier PR #7593 has already been merged,

Yes...   I understand.   But I think what is going to happen is that Pull Request will get re-based against the v2.0.0 code when it shows up in the next week.   I'm not positive about this, but I think Thinky wants bugfix-v1.1.0 to stay frozen until that happens.

**Comment by marcio-ao** [comment]
@Roxy-3D: Ah. Okay. I'll be patient then :)

**Comment by Roxy-3D** [comment]
Don't worry...   It will happen!   But we are in kind of an awkward transition phase right now.

**Comment by cdedwards** [comment]
I agree the code should be shared.. when I attempted that however, it was not finding the graphic char arrays, so I went a different route for now. The overlay function, and the graphics should be made slightly more generic and then can be used in several different places where the nozzle and bed are changing, UBL is just one example. When 2.0 is released I will take another stab at this, or leave it in your capable hands.

**Comment by Roxy-3D** [comment]
> When 2.0 is released I will take another stab at this, or leave it in your capable hands.

We can work on this now...  And when v2.0.0 shows up...  It can be the first merge to the new branch.
And in fact...   I would argue...  We can just keep modifying this Pull Request to make it 'correct' with everybody agreeing to the feature set.

Everybody (myself included) is learning from this.   Let's just plow ahead and make sure everybody's ideas are addressed.

**Comment by marcio-ao** [comment]
@cdedwards: I think you are on the right track. The reason you probably were not finding the graphics is that you need to replace "#if ENABLED(BABYSTEP_ZPROBE_GFX_OVERLAY)" with "#if ENABLED(BABYSTEP_ZPROBE_GFX_OVERLAY) || ENABLED(AUTO_BED_LEVELING_UBL)" in "Marlin/dogm_bitmaps.h" so that way the graphics will be loaded if either BABYSTEP_ZPROBE_GFX_OVERLAY *or* AUTO_BED_LEVELING_UBL are defined.

**Comment by marcio-ao** [comment]
@roxy-3d: Not sure why the nozzle graphics isn't showing up for you. That remains a mystery to me. But the line in the code that draws that bitmap is the following:

 if(PAGE_CONTAINS(3,16))  u8g.drawBitmapP(nozzle + 6, 4 - dir,2,12,nozzle_ubl_bmp);

You could try a couple things. Replace that line with:

u8g.drawBitmapP(nozzle + 6, 4 - dir,2,12,nozzle_ubl_bmp);

The PAGE_CONTAINS(...) is an optimization, but you can see if it works without it. If that doesn't work, try this:

u8g.drawBitmapP(nozzle + 6, 4,2,12,nozzle_ubl_bmp);

The "dir" variable is what nudges the nozzle up and down. The value should be either 1 or -1, depending on which direction the value is chaning, but if for some reasons it is not, the above test will reveal it (by drawing the graphics at a fixed location of 4).

**Comment by marcio-ao** [comment]
@Roxy-3D: The following page explains how graphics are drawn and was helpful to me in understanding how Marlin drawing graphics:

https://github.com/olikraus/u8glib/wiki/tpictureloop

It turns out that draw function is actually called multiple times, in a loop, for different "pages" of the screen. The PAGE_CONTAINS(y1,y2) macro checks to see whether Marlin is drawing rows y1 through y2. The arguments to are u8g.drawBitmapP(x,y,bytes_across,rows_down,bmp), so the following line:

if(PAGE_CONTAINS(3,16)) u8g.drawBitmapP(nozzle + 6, 4 - dir,2,12,nozzle_ubl_bmp);

Means:

If rows 3 through 16 are in the page which is being drawn, then draw the bitmap 2 bytes wide and 12 rows tall at (x,y) = (nozzle+6, 4 - dir).

The fact that you are seeing a dot at the top of the screen makes me suspect that the bitmap isn't being drawn at the correct location. You can try modifying the first two parameters of drawBitmapP to see if you can get it to show up.

**Comment by Tannoo** [comment]
@Roxy-3D, I have not looked at this yet. Been busy with life atm. I want to apply this to the 2.0.0 and try it. It sounds like a wonderful addition. Still a few days from now.

**Comment by Tannoo** [comment]
>I would suggest merging PR #7602 as my earlier PR #7593 has already been merged...

I have included these two PR's into my bugfix-2.0.x and had some errors about the bitmaps not being defined. Well, I defined them and now have wierd lines as if the bitmaps are not interlaced correctly.

I am trying to get this to work before looking into the 20x4.

**Comment by cdedwards** [comment]
I'm attempting to implement this on the refactor version of 2.0. currently stuck at 

linking .pioenvs\Re-ARM\firmware.elf
.pioenvs/Re-ARM/src/src/lcd/ultralcd.o: In function `U8GLIB::U8GLIB(_u8g_dev_t*, unsigned char, unsigned char, unsigned char)':
I:\GitHub\Marlin/.piolibdeps\U8glib-ARM_ID1691\src/U8glib.h:73: undefined reference to `U8GLIB::initHWSPI(_u8g_dev_t*, unsigned char, unsigned char, unsigned char)'
collect2.exe: error: ld returned 1 exit status
*** [.pioenvs\Re-ARM\firmware.elf] Error 1
 [ERROR] Took 13.19 seconds

I've had to specifically enable DOGLCD in my configuration.h.  As far as I can see this is defined in u8glib.h

 /* constructor */
    U8GLIB(void)
      { }
    U8GLIB(u8g_dev_t *dev)
      { prepare(); u8g_Init(&u8g, dev); }
    U8GLIB(u8g_dev_t *dev, u8g_com_fnptr com_fn)
      { prepare(); u8g_InitComFn(&u8g, dev, com_fn); }
    U8GLIB(u8g_dev_t *dev, uint8_t sck, uint8_t mosi, uint8_t cs, uint8_t a0, uint8_t reset) 
      { initSPI(dev, sck, mosi, cs, a0, reset); }
    U8GLIB(u8g_dev_t *dev, uint8_t cs, uint8_t a0, uint8_t reset) 
      { initHWSPI(dev, cs, a0, reset); }

**Comment by cdedwards** [comment]
as for the missing char maps, need to move the HAS_TEMP_BED in the dogm_bitmaps.h. specifially the #endif 

that was what was causing the issue for my attempt at adding this to ubl.

**Comment by Tannoo** [comment]
Is this PR updated with the latest that works?

This PR doesn't have any modifications to the `dogm_bitmaps.h`.

**Comment by cdedwards** [comment]
no it isn't. since it's possible to combine the two so there is only one set of functions.

**Pull request closed without merging**